### PR TITLE
Feature/24797 integrationsdata

### PR DESCRIPTION
**Pull request by jgj-magenta-aps**
- [X] Have you updated the documentation?
- [X] Have you performed a smoke test of the application?
- [X] Have you written tests for your changes?
- [X] Have you updated the release notes?

### CODE REVIEW

**Comment by jgj-magenta-aps** [comment]
Oh Yes - there's a new db_structure coming in - and it has to walk from MO-package and into oio_common by itself.

File: README.rst [hunk 1]
```diff
@@ -53,6 +53,15 @@ En `LoRa <https://github.com/magenta-aps/mox>`_ backend, som gemmer alle data
...
 RESTful service skrevet i Python. LoRa kan opfattes som *Model* i MVC-modellen.
 
+MO betjener sig af tilretninger af datamodellen i LoRa. Før Lora kan anvendes sammen 
+med MO skal disse tilretninger afspejles i databasen. 
+
+MO-tilretninger af datamodellen i LoRa
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+For at får datamodellen i LoRa til at afspejle datamodellen i MO skal man, 
+inden systemet tages i brug, kopiere MO's: *os2mo/setup/db_structure.py* til
+Loras: *oio_rest/oio_common/db_structure.py* og afvikle recreatedb.sh
```
[the remark this answers is not in the document]
Comment (agger-magenta) [reply README.rst]: I øvrigt tror jeg, sætningen vil fungere bedre uden kommaerne omkring "inden systemet tages i brug".

File: backend/mora/common.py [hunk 2]
```diff
@@ -374,6 +374,7 @@ def create_organisationsenhed_payload(
     enhedstype: str,
     overordnet: str,
     adresser: typing.List[dict] = None,
+    integrationdata: str = ""
```
Comment (agger-magenta) [inline backend/mora/common.py]: Since these are two words, it would be better to call this variable 'integration_data`.

File: backend/mora/cli/__init__.py [hunk 3]
```diff
@@ -6,5 +6,29 @@
 # file, You can obtain one at http://mozilla.org/MPL/2.0/.
 #
 
+import os
+import sys
+
+# cant do monkeypatching only because file is used as program
```
Comment (agger-magenta) [inline backend/mora/cli/__init__.py]: "cant" -> "Can't"

File: backend/mora/common.py [hunk 4]
```diff
@@ -427,7 +437,8 @@ def create_bruger_payload(
     brugernavn: str,
     brugervendtnoegle: str,
     tilhoerer: str,
-    cpr: str
+    cpr: str,
+    integrationdata: str,
```
Comment (agger-magenta) [inline backend/mora/common.py]: Also here: `integration_data`

File: backend/mora/mapping.py [hunk 5]
```diff
@@ -34,6 +34,7 @@
 LOCATION = 'location'
 ERROR = 'error'
 USER_SETTINGS = 'user_settings'
+INTEGRATIONDATA = 'integrationdata'
```
Comment (agger-magenta) [inline backend/mora/mapping.py]: Igen, et `_` ville være bedre.

File: backend/mora/common.py (hunk 2 above)
Comment (danchr) [reply backend/mora/common.py]: It seems to me that using Danish would be consistent with the other parameters, so `integrationsdata`.

File: backend/mora/mapping.py (hunk 5 above)
Comment (danchr) [reply backend/mora/mapping.py]: In this case, I agree — the other variables are in English :)

File: backend/mora/common.py (hunk 2 above)
Comment (agger-magenta) [reply backend/mora/common.py]: Well, but if it's supposed to be read as English, it should be with an underscore. :-)

File: backend/mora/cli/__init__.py (hunk 3 above)
Comment (danchr) [reply backend/mora/cli/__init__.py]: > "cant" -> "Can't"

Alternatively, [cannot](https://en.oxforddictionaries.com/usage/cannot-or-can-not) 🙂

File: backend/mora/common.py [hunk 6]
```diff
@@ -416,6 +417,15 @@ def create_organisationsenhed_payload(
     if adresser:
         org_unit['relationer']['adresser'] = adresser
 
+    if integration_data:
+        org_unit[
```
Comment (agger-magenta) [inline backend/mora/common.py]: Hvorfor behandle integrationsdata anderledes end de andre egenskaber?

Comment (agger-magenta) [reply backend/mora/common.py]: Altså, de kunne også komme med sammen med enhedsnavn og brugervendtnøgle. Så vil de blot i nogle tilfælde være den tomme streng, men de vil altid være der (i MO, altså).

File: backend/mora/common.py [hunk 7]
```diff
@@ -464,6 +475,15 @@ def create_bruger_payload(
             },
         ]
 
+    if integration_data:
```
Comment (agger-magenta) [inline backend/mora/common.py]: Samme spørgsmål som før.

File: backend/mora/service/orgunit.py [hunk 8]
```diff
@@ -326,8 +342,12 @@ def get_one_orgunit(c, unitid, unit=None,
             rels['tilhoerer'][0]['uuid'],
         )
 
+    elif details is UnitDetails.MINIMAL:
+        pass  # already done
+    elif details is UnitDetails.INTEGRATION:
+        r["integration_data"] = attrs.get("integrationsdata", "")
```
Comment (agger-magenta) [inline backend/mora/service/orgunit.py]: eller det her.

File: backend/mora/service/employee.py [hunk 9]
```diff
@@ -199,7 +224,10 @@ def get_one_employee(c, userid, user=None, full=False):
 
         r[mapping.ORG] = org.get_one_organisation(c, orgid)
         r[mapping.USER_KEY] = props['brugervendtnoegle']
-
+    elif details is EmployeeDetails.MINIMAL:
+        pass  # already done
+    elif details is EmployeeDetails.INTEGRATION:
+        r["integration_data"] = props.get("integrationsdata", "")
```
Comment (agger-magenta) [inline backend/mora/service/employee.py]: ... for hvis du ikke gav integrationsdata særbehandling herover, ville du ikke behøve at gøre det her.

File: NEWS.rst [hunk 10]
```diff
@@ -6,6 +6,7 @@ Internal changes
 
 * 23799: Added custom UUID url converter, stringifying UUID parameters in
   order to standardise our use of UUIDs internally.
+* #24797 Integration_data added to employee and org.unit
```
Comment (agger-magenta) [inline NEWS.rst]: Strengt taget er dette her vel almindelig tale, og så vil det være mest naturligt at skrive ("integration data") i to ord uden underscore (undskyld, jeg er så vanskelig! :-))

File: README.rst [hunk 11]
```diff
@@ -226,6 +235,10 @@ feature kræver det følgende afhængigheder::
 
   $ sudo apt install libxmlsec1-dev libxmlsec1-openssl postgresql-contrib
 
+For at få datamodellen i Loras testinstans til at afspejle datamodellen i MO
```
Comment (danchr) [inline README.rst]:
```suggestion
For at få datamodellen i LoRas testinstans til at afspejle datamodellen i MO 
```

File: README.rst [hunk 12]
```diff
@@ -226,6 +235,10 @@ feature kræver det følgende afhængigheder::
 
   $ sudo apt install libxmlsec1-dev libxmlsec1-openssl postgresql-contrib
 
+For at få datamodellen i Loras testinstans til at afspejle datamodellen i MO 
+vil MO's: *os2mo/setup/db_structure.py* overskrive oio_rests: *oio_common/db_structure.py*
```
Comment (danchr) [inline README.rst]:
```suggestion
vil MOs ``os2mo/setup/db_structure.py`` overskrive oio_rests ``oio_common/db_structure.py``
```

File: README.rst [hunk 13]
```diff
@@ -226,6 +235,10 @@ feature kræver det følgende afhængigheder::
...
+i det aktuelle python-environment
```
Comment (danchr) [inline README.rst]:
```suggestion
i det aktuelle python-environment.
```

File: backend/mora/cli/__init__.py [hunk 14]
```diff
@@ -6,5 +6,29 @@
...
+    import oio_common
+    from oio_common import db_structure
+    orgfile = db_structure.__file__
+    backenddir = os.path.dirname(os.path.dirname(os.path.dirname(__file__)))
+    topdir = os.path.dirname(backenddir)
+    patchfile = os.path.join(topdir, 'setup', "db_structure.py")
+    if not open(orgfile).read() == open(patchfile).read():
+        open(orgfile, "w").write(open(patchfile).read())
+        # monkeypatching necessary for
+        # first time running tests (jenkins)
```
Comment (danchr) [inline backend/mora/cli/__init__.py]: 👎 — on general principle, test logic should reside in `tests`, and nowhere else. Furthermore, this file is not the proper location for this logic; it's just too magic to have this occur as a side-effect of importing `mora.cli` package.

When do we need to do this? If only during tests, we should do this in the `setUp` or `setUpClass` methods on the base class. (Not only will this make the tests more self-contained, it'll also ensure that the tests don't _rely_ on being called by our CLI.)

File: backend/mora/service/employee.py [hunk 15]
```diff
@@ -317,7 +345,18 @@ def get_employee(id):
     '''
     c = common.get_connector()
 
-    r = get_one_employee(c, id, full=True)
+    details = flask.request.args.get("employeedetails", EmployeeDetails.FULL)
```
Comment (danchr) [inline backend/mora/service/employee.py]: The documentation should mention this parameter. The fact that you're acting on an employee is implied by the endpoint; hence including the string `employee` in the argument name is redundant. Furthermore, if it must be optional, I'd prefer a dedicated boolean parameter, such as `integregationdata=1` — see `utils.get_args_flag()`.

What does generating this data involve? If next to nothing, we might as well include this by default and drop the parameter, as @agger-magenta suggests.

File: backend/mora/service/orgunit.py [hunk 16]
```diff
@@ -477,7 +497,18 @@ def get_orgunit(unitid):
     '''
     c = common.get_connector()
 
-    r = get_one_orgunit(c, unitid, details=UnitDetails.FULL)
+    details = flask.request.args.get("unitdetails", UnitDetails.FULL)
```
Comment (danchr) [inline backend/mora/service/orgunit.py]: Needs documentation if retained, and should be called the same as the relevant argument to `/e/`.

File: backend/tests/fixtures/create_bruger_andersine.json [hunk 17]
```diff
@@ -4,6 +4,7 @@
             {
                 "brugernavn": "Andersine And",
                 "brugervendtnoegle": "andersineand",
+                "integrationsdata": "{\"von-and-løn-id\": \"2468\"}",
```
Comment (danchr) [inline backend/tests/fixtures/create_bruger_andersine.json]: Not your fault, but JSON-in-JSON strikes me as quite horrible 😞

File: backend/tests/fixtures/create_organisationenhed_hum.json [hunk 18]
```diff
@@ -4,6 +4,7 @@
             {
                 "brugervendtnoegle": "hum",
                 "enhedsnavn": "Humanistisk fakultet",
+                "integrationsdata":"{}",
```
Comment (danchr) [inline backend/tests/fixtures/create_organisationenhed_hum.json]: You should also test a `null` value — i.e. simply not specifying it.

File: backend/tests/test_integration_employee.py [hunk 19]
```diff
@@ -679,3 +679,112 @@ def test_edit_employee(self):
...
+        self.load_sample_structures()
+        employee_uuid = 'df55a3ad-b996-4ae0-b6ea-a3241c4cbb24'
+        util.load_fixture('organisation/bruger',
+                          'create_bruger_andersine.json',
+                          employee_uuid)
+
+        self.assertRequestResponse(
+            '/service/e/{}/?employeedetails=INTEGRATION'.format(employee_uuid),
+            {
+                'integration_data': '{"von-and-løn-id": "2468"}',
```
Comment (danchr) [inline backend/tests/test_integration_employee.py]: There's no reason why we can't unparse the JSON, is there?

File: backend/tests/test_integration_org_unit.py [hunk 20]
```diff
@@ -918,8 +920,14 @@ def test_edit_org_unit(self):
 
         self.assertRegistrationsEqual(expected, actual)
 
+    @unittest.expectedFailure
```
Comment (danchr) [inline backend/tests/test_integration_org_unit.py]: Not a fan of this. Can we edit the tests to still pass? History shows that once a test is marked `skip` or `expectedFailure`, no-one will care or notice that it's wrong 😕

**Comment by danchr** [comment]
Please note that almost all tests fail with an error like this:

```Traceback (most recent call last):
  File "/src/mora/PR-216-head/venv/lib/python3.5/site-packages/requests_mock/mocker.py", line 225, in inner
    return func(*args, **kwargs)
  File "/src/mora/PR-216-head/backend/tests/test_integration_address.py", line 2794, in test_missing_address
    self.load_sample_structures()
  File "/src/mora/PR-216-head/backend/tests/util.py", line 507, in load_sample_structures
    load_sample_structures(**kwargs)
  File "/src/mora/PR-216-head/backend/tests/util.py", line 244, in load_sample_structures
    load_fixture(path, fixture_name, uuid)
  File "/src/mora/PR-216-head/backend/tests/util.py", line 96, in load_fixture
    r = lora.create(path, get_fixture(fixture_name, **kwargs), uuid)
  File "/src/mora/PR-216-head/backend/mora/lora.py", line 209, in create
    _check_response(r)
  File "/src/mora/PR-216-head/backend/mora/lora.py", line 159, in _check_response
    message=msg, cause=cause)
mora.exceptions.HTTPException: 400 Bad Request: malformed record literal: "{}"
```

File: backend/mora/service/employee.py (hunk 15 above)
Comment (agger-magenta) [reply backend/mora/service/employee.py]: :+1:

File: backend/tests/fixtures/create_bruger_andersine.json (hunk 17 above)
Comment (agger-magenta) [reply backend/tests/fixtures/create_bruger_andersine.json]: As Dan writes below, there's not actually any reason to have the JSON escaped. It can be nested so that it's treated completely like nested JSON, even if MO is doing nothing but carry it around. Nothing is gained by having it as a string, as far as I can see.

Or is there some complication I haven't thought of?

File: backend/mora/common.py (hunk 6 above)
Comment (jgj-magenta-aps) [reply backend/mora/common.py]: Det kan du bestemt have ret i - Jeg så bare adressers speccialbehandling og så tænkte jeg: Måske er den der og måske ikke, men det tror jeg uden at have kigget videre, at jeg sagtens kan give dig ret i - den skal bare defaulte til den tomme streng.
-
Ved at bruge en streng, kan vi i øvrigt være agnostiske - hvis en implementør vil anvende xml i integrationsdata, fordi det passer bedre, så er det vel også i orden?

File: backend/mora/common.py (hunk 7 above)
Comment (jgj-magenta-aps) [reply backend/mora/common.py]: jeg er enig - men stadig med den holdning at vi skal lagre en streng i stedet for et objekt

File: backend/mora/service/orgunit.py (hunk 8 above)
Comment (jgj-magenta-aps) [reply backend/mora/service/orgunit.py]: Her er meningen at man kun henter integrationsdata ud, når man skal bruge dem - udfra devisen, at der faktisk kunne likke 10 k data for hver organisation, og dem vil vi ikke have med op i brugerinterfacet.

File: backend/mora/common.py (hunk 6 above)
Comment (agger-magenta) [reply backend/mora/common.py]: Re: Streng har du en pointe. I princippet kunne integratøren så sende et UU-encoded JPG-billede med dataene gemt steganorafisk, og MO ville stadig være ligeglad. :dolphin:

File: backend/mora/service/employee.py (hunk 9 above)
Comment (jgj-magenta-aps) [reply backend/mora/service/employee.py]: Samme kommentar som ovenfor: Dette er udelukkende for at brugerinterfacet ikke skal gøre noget for at undgå integrationsdata, men de, som vil anvende dem - integrationsprogrammer - skal specifikt bede om at se dem.

File: backend/mora/service/orgunit.py (hunk 8 above)
Comment (agger-magenta) [reply backend/mora/service/orgunit.py]: Det er også fint. Faktisk bør integrationsdata aldrig sendes netop til brugerinterfacet, gætter jeg på.

Det vil sige, man skal kun have dem, når integrationsparametren er sat. Men det min kommentar referer til er, at hvis du sætter `integrationsdata` ligesom du sætter brugervendt nøgle osv., så kan du skrive `attrs['integrationsdata']` i stedet for `attrs.get(...)`

File: backend/mora/service/employee.py (hunk 9 above)
Comment (agger-magenta) [reply backend/mora/service/employee.py]: Samme kommentar som ovenfor - det er ikke if-sætningen, jeg brokker mig over :slightly_smiling_face:

File: backend/mora/service/orgunit.py (hunk 8 above)
Comment (jgj-magenta-aps) [reply backend/mora/service/orgunit.py]: Jeg er jo ikke lorekspert, men kan man forestille sig, at der i en post (fra en ældre lora-db) ikke er integrationsdata, men at det så tilføjes til db_structure.py uden at man retter alle poster?
-
Og hvis det ske, ville vi så ikke have en props uden? 
-
Kan det i givet fald ske i fremtiden, når vi ændrer i databasen?
I db_structures.py ser det ud til at det kun er brugervendtnøgle, der er mandatory ifølge metadata.
Betyder det ikke at der måske ikke er integrationsdata?

File: backend/mora/service/employee.py (hunk 9 above)
Comment (jgj-magenta-aps) [reply backend/mora/service/employee.py]: Jeg har forfægtet min forsigtighed ovenfor.

File: backend/mora/service/orgunit.py (hunk 8 above)
Comment (agger-magenta) [reply backend/mora/service/orgunit.py]: Nej, man kunne fra nu af ikke forestille sig, at MO-versioner med et integrationsdata-felt i egenskaberne kørte op mod en LoRa, som ikke har det. Det giver ikke mening.

Comment (agger-magenta) [reply backend/mora/service/orgunit.py]: Vores AAK-integration kan jo heller ikke køre op mod en LoRa, som ikke har vores særlige udvidelsesfelter.

Comment (agger-magenta) [reply backend/mora/service/orgunit.py]: Dvs. jeg tænker, at hvis en egenskab ikke er mandatory og mangler, skal det tolkes som, at den er None eller null eller tom eller whatever, ikke som at den slet ikke er der.

File: NEWS.rst (hunk 10 above)
Comment (jgj-magenta-aps) [reply NEWS.rst]: Er delt!

File: README.rst (hunk 11 above)
Comment (jgj-magenta-aps) [reply README.rst]: Skulle være fixet

File: README.rst (hunk 12 above)
Comment (jgj-magenta-aps) [reply README.rst]: Er rettet

File: README.rst (hunk 13 above)
Comment (jgj-magenta-aps) [reply README.rst]: rettet!

File: backend/mora/cli/__init__.py (hunk 14 above)
Comment (jgj-magenta-aps) [reply backend/mora/cli/__init__.py]: I eneded up asking Jo, where she'd put it - I wanted to keep it out of setUpClass, as we don't inherit from the same class in all tests.
Jo suggested __init__.py in tests, and I went with it - it seems to avoid import during flask run, and that was, I guess, the main issue.

**Comment by jgj-magenta-aps** [comment]
Jeg har set lidt på sagerne, men jeg tør ikke at rette i de test-cafe's, der fejler

File: backend/mora/cli/__init__.py (hunk 14 above)
Comment (danchr) [reply backend/mora/cli/__init__.py]: > I eneded up asking Jo, where she'd put it - I wanted to keep it out of setUpClass, as we don't inherit from the same class in all tests.
> Jo suggested **init**.py in tests, and I went with it - it seems to avoid import during flask run, and that was, I guess, the main issue.

All tests that access LoRA descend from the same test class and/or mixin that grants them access to the database. To allow sharing it between LoRA and os2mo, the actual implementation currently exists in `oio_rest.utils.test_support` — but we have subclasses in `os2mo` where you can add an extra method.

I respectfully disagree with Io in this regard — having code of this order in `__init__.py` is simply too implicit and magical.

File: backend/mora/service/orgunit.py [hunk 21]
```diff
@@ -57,6 +58,9 @@ class UnitDetails(enum.Enum):
     # same as above, but with all parents
     FULL = 3
 
+    # minimal and integration_data
+    INTEGRATION = 4
```
[the remark this answers is not in the document]
Comment (jgj-magenta-aps) [reply backend/mora/service/orgunit.py]: Actually the FULL line is Yours (merged in), but I can take it away, sure.

File: backend/mora/cli/base.py [hunk 22]
```diff
@@ -150,6 +150,7 @@ def test(tests, quiet, verbose, minimox_dir, browser, do_list,
          keywords, xml_report, **kwargs):
     '''Test the application.'''
     sys.path.insert(0, backenddir)
+    import tests as _
```
[the remark this answers is not in the document]
Comment (jgj-magenta-aps) [reply backend/mora/cli/base.py]: Yup - probably a slip from testing.

File: backend/tests/fixtures/create_bruger_andersine.json (hunk 17 above)
Comment (jgj-magenta-aps) [reply backend/tests/fixtures/create_bruger_andersine.json]: Yes - Json fields was a nogo - but convrint on the way in and out was.

File: backend/mora/common.py (hunk 6 above)
Comment (jgj-magenta-aps) [reply backend/mora/common.py]: Nu er det blevet JSON, Carsten - ind og ud - men det er stadig en streng i Lora.

File: backend/tests/test_integration_org_unit.py (hunk 20 above)
Comment (jgj-magenta-aps) [reply backend/tests/test_integration_org_unit.py]: Removed 2 expected failures

File: backend/mora/service/orgunit.py (hunk 21 above)
[the remark this answers is not in the document]
Comment (jgj-magenta-aps) [reply backend/mora/service/orgunit.py]: I haven't gotten around to changing this yet,but I will create extra endpoints for this instead. Will however let the enum stay both places, as it seems like a good way to control behaviour

File: README.rst [hunk 23]
```diff
@@ -54,6 +54,15 @@ i en PostgreSQL-database. Disse data udstilles og manipuleres via en
...
 LoRa anvender OIO-standarderne for sag, dokument, organisation og klassifikation
 
+MO betjener sig af tilretninger af datamodellen i LoRa. Før Lora kan anvendes sammen 
+med MO skal disse tilretninger afspejles i databasen. 
+
+MO-tilretninger af datamodellen i LoRa
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+For at få datamodellen i LoRa til at afspejle datamodellen i MO skal man 
+inden systemet tages i brug kopiere MO's: os2mo/setup/db_structure.py* til
+Loras: *oio_rest/oio_common/db_structure.py* og afvikle recreatedb.sh
```
Comment (danchr) [inline README.rst]:
```suggestion
til ``mox/oio_rest/oio_common/db_structure.py`` og afvikle ``recreatedb.sh``
```

File: README.rst [hunk 24]
```diff
@@ -54,6 +54,15 @@ i en PostgreSQL-database. Disse data udstilles og manipuleres via en
 RESTful service skrevet i Python. LoRa kan opfattes som *Model* i MVC-modellen.
 LoRa anvender OIO-standarderne for sag, dokument, organisation og klassifikation
 
+MO betjener sig af tilretninger af datamodellen i LoRa. Før Lora kan anvendes sammen 
+med MO skal disse tilretninger afspejles i databasen. 
+
+MO-tilretninger af datamodellen i LoRa
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+For at få datamodellen i LoRa til at afspejle datamodellen i MO skal man 
+inden systemet tages i brug kopiere MO's: os2mo/setup/db_structure.py* til
```
Comment (danchr) [inline README.rst]:
```suggestion
inden systemet tages i brug kopiere ``os2mo/setup/db_structure.py`` til
```

File: README.rst [hunk 25]
```diff
@@ -226,6 +235,10 @@ feature kræver det følgende afhængigheder::
 
   $ sudo apt install libxmlsec1-dev libxmlsec1-openssl postgresql-contrib
 
+For at få datamodellen i LoRas testinstans til at afspejle datamodellen i MO 
+vil MO's: ``os2mo/setup/db_structure.py`` overskrive oio_rests: ``oio_common/db_structure.py``
```
Comment (danchr) [inline README.rst]:
```suggestion
vil MOs ``os2mo/setup/db_structure.py`` overskrive oio_rests ``oio_common/db_structure.py``
```

File: README.rst [hunk 26]
```diff
@@ -226,6 +235,10 @@ feature kræver det følgende afhængigheder::
...
+i det aktuelle python-environment.
```
Comment (danchr) [inline README.rst]:
```suggestion
i det aktuelle pythonmiljø.
```

I øvrigt er det forhåbentligt ikke tilfældet længere — jeg har tilrettet `test_support.py` så det ikke burde være nødvendigt. Så længe denne ændring overskriver kildekode i pythonmiljøet får den altså et :-1: herfra.

File: backend/mora/service/integration_data.py [hunk 27]
```diff
@@ -0,0 +1,102 @@
...
+    .. :quickref: Unit ; integration data
+
+    :param uuid unitid: UUID of the unit to retrieve.
+
+    :queryparam date at: Show the unit at this point in time,
+        in ISO-8601 format.
+
+    :statuscode 200: Integration data (json) retrieved
+    :statuscode 404: No such unit found.
+
```
Comment (danchr) [inline backend/mora/service/integration_data.py]: Please include an example response.

File: backend/mora/service/employee.py [hunk 28]
```diff
@@ -191,7 +219,10 @@ def get_one_employee(c, userid, user=None, full=False):
 
         r[mapping.ORG] = org.get_one_organisation(c, orgid)
         r[mapping.USER_KEY] = props['brugervendtnoegle']
-
+    elif details is EmployeeDetails.MINIMAL:
+        pass  # already done
+    elif details is EmployeeDetails.INTEGRATION:
+        r[mapping.INTEGRATION_DATA] = props.get("integrationsdata")
```
Comment (danchr) [inline backend/mora/service/employee.py]: We should parse the JSON here. Do we have tests in place for the case where someone injects invalid JSON into the attribute, or where it's unset?

File: backend/mora/service/integration_data.py [hunk 29]
```diff
@@ -0,0 +1,102 @@
...
+    """
+    c = common.get_connector()
+
+    r = orgunit.get_one_orgunit(c, unitid,
+                                details=orgunit.UnitDetails.INTEGRATION)
+
+    if not r:
+        exceptions.ErrorCodes.E_ORG_UNIT_NOT_FOUND(org_unit_uuid=unitid)
+
+    if r[mapping.INTEGRATION_DATA]:
```
Comment (danchr) [inline backend/mora/service/integration_data.py]: Wouldn't it be nicer if this simply returned the contents of the integration data? (Also, parsing JSON should happen where it's extracted, as mentioned earlier.)

File: backend/mora/service/integration_data.py [hunk 30]
```diff
@@ -0,0 +1,102 @@
...
+    """
+    c = common.get_connector()
+
+    r = employee.get_one_employee(c, employeeid,
+                                  details=employee.EmployeeDetails.INTEGRATION)
+
+    if not r:
+        exceptions.ErrorCodes.E_USER_NOT_FOUND(employee_uuid=employeeid)
+
+    if r[mapping.INTEGRATION_DATA]:
```
Comment (danchr) [inline backend/mora/service/integration_data.py]: Likewise: 

> Wouldn't it be nicer if this simply returned the contents of the integration data? (Also, parsing JSON should happen where it's extracted, as mentioned earlier.)

File: backend/tests/lorapatch.py [hunk 31]
```diff
@@ -0,0 +1,34 @@
...
+    try:
+        # patch db_structure in case we have oio_rest/oio_common installed
+        import oio_common  # noqa F401
+        from oio_common import db_structure
+        orgfile = pathlib.Path(db_structure.__file__)
+        topdir = pathlib.Path(__file__).absolute().parent.parent.parent
+        patchfile = topdir / "setup" / "db_structure.py"
+
+        if orgfile.read_text() != patchfile.read_text():
+            orgfile.write_text(patchfile.read_text())
```
Comment (danchr) [inline backend/tests/lorapatch.py]: This does not belong in a separate module, but in `utils.py`, and should most certainly not happen as a side-effect of importing said module. Overwriting a Python source file is evil enough as-is, but doing so at import-time is even worse.

Still: You shouldn't need this. If you do, let's talk about how to adjust LoRA so you don't.

File: backend/tests/test_integration_org_unit.py [hunk 32]
```diff
@@ -1570,7 +1583,12 @@ def test_rename_org_unit(self):
...
```
[the remark this answers is not in the document]
Comment (danchr) [reply backend/tests/test_integration_org_unit.py]: A new `@expectedFailure`? :)

File: backend/tests/util.py [hunk 33]
```diff
@@ -26,11 +27,14 @@
 import time
 import werkzeug.serving
 
+from .lorapatch import LORA_PATCHED
```
Comment (danchr) [inline backend/tests/util.py]: :-1:

File: README.rst (hunk 26 above)
Comment (jgj-magenta-aps) [reply README.rst]: Jeg fik ikke dine ændringer desangående til at fungere for mig - husk - det eker på jenkins

File: backend/mora/service/employee.py (hunk 28 above)
Comment (jgj-magenta-aps) [reply backend/mora/service/employee.py]: Maybe and no- we probably do not have that, But then again - it is not json that is injected - it is an object that is encoded using json.dumps - see above:

            if mapping.INTEGRATION_DATA in data:
                attrs['integrationsdata'] = json.dumps(
                    data[mapping.INTEGRATION_DATA]
                )

File: backend/mora/service/integration_data.py (hunk 27 above)
Comment (jgj-magenta-aps) [reply backend/mora/service/integration_data.py]: Will do, certainly.

File: backend/mora/service/integration_data.py (hunk 29 above)
Comment (jgj-magenta-aps) [reply backend/mora/service/integration_data.py]: I am not sure what You mean here?

File: backend/mora/service/integration_data.py (hunk 30 above)
Comment (jgj-magenta-aps) [reply backend/mora/service/integration_data.py]: Pretty soon there will be integration data in all objects, and then we may have to change this.

File: backend/tests/lorapatch.py (hunk 31 above)
Comment (jgj-magenta-aps) [reply backend/tests/lorapatch.py]: This will have a short life. Pretty soon there will be integration data in all objects in lora. Using copious amounts of hours to fix something that will be gone in a split second seems a little over the top. But if You can get it to work without this, be my guest.

File: backend/mora/service/employee.py (hunk 28 above)
Comment (danchr) [reply backend/mora/service/employee.py]: > Maybe and no- we probably do not have that, But then again - it is not json that is injected - it is an object that is encoded using json.dumps - see above:

AFAICT, no such code exists within this function. `props` is assigned `user['attributter']['brugeregenskaber'][0]` in line 204, and not modified after that. In this particular statement, you're assigning the raw, unmodified value of the LoRA attribute — i.e. a string. You should assign the parsed JSON instead:
```suggestion
        r[mapping.INTEGRATION_DATA] = json.parse(props.get("integrationsdata", "null"))
```
…or something similar, depending on how you want to handle errors.

File: backend/tests/test_integration_org_unit.py (hunk 32 above)
[the remark this answers is not in the document]
Comment (jgj-magenta-aps) [reply backend/tests/test_integration_org_unit.py]: Nope - one I had prematurely removed on commit 659ca54, reinstated as it seems non related.

File: README.rst (hunk 26 above)
Comment (jgj-magenta-aps) [reply README.rst]: I respektfully disagree on using 'pythonmiljø' - this word only gets 3 hits on goole. 
Python-environment should be the preferred term in danish too.

File: backend/mora/service/integration_data.py (hunk 29 above)
Comment (jgj-magenta-aps) [reply backend/mora/service/integration_data.py]: Well it was You who wanted the integration data handling moved out of the orgunit/employee endpoints. 
They have been moved now, and the behavoiur is symmetrical: 
They are encoded into json upon entering mo end decoded upon exiting mo.

File: backend/tests/util.py (hunk 33 above)
Comment (jgj-magenta-aps) [reply backend/tests/util.py]: This will have a short life. Pretty soon there will be integration data in all objects in lora. Using copious amounts of hours to fix something that will be gone in a split second seems a little over the top. But if You can get it to work without this, be my guest.

File: README.rst (hunk 26 above)
Comment (agger-magenta) [reply README.rst]: Det hedder "pythonmiljø", hvis det hedder "udviklingsmiljø" - og "python-environment", hvis det hedder "udviklings-environment".

Det er altså det første, der er rigtigt.

File: backend/tests/util.py (hunk 33 above)
Comment (danchr) [reply backend/tests/util.py]: I believe we'll eventually want this going forward, so I'll take a look!

File: backend/mora/service/integration_data.py (hunk 27 above)
Comment (jgj-magenta-aps) [reply backend/mora/service/integration_data.py]: Examples added.
